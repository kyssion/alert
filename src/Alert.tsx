import React, { useState } from 'react';
import { createGlobalStyle } from 'styled-components';
import { CSSTransition } from 'react-transition-group';
import {
  AiFillCheckCircle,
  AiOutlineCheckCircle,
  AiOutlineInfoCircle,
  AiFillInfoCircle,
  AiFillCloseCircle,
  AiFillExclamationCircle,
  AiOutlineExclamationCircle,
  AiOutlineCloseCircle,
  AiOutlineClose,
} from 'react-icons/ai';
import classNames from 'classnames';
import assert from 'assert';
import AlertMessage from './AlertMessage';
import AlertDescription from './AlertDescription';
import AlertLayout from './AlertLayout';
import DenseIcon from './DenseIcon';
import CloseButton from './CloseButton';

const timeout = 200;

const Globalstyle = createGlobalStyle`
  .sinoui-alert-exit {
    opacity: 1;
  }
  .sinoui-alert-exit-active {
    opacity: 0;
    transform: scale(0.9);
    height: 0!important;
    transition: opacity ${timeout}ms, transform ${timeout}ms, height ${timeout}ms;
  }
`;

interface Props {
  /**
   * 提示内容
   */
  message?: React.ReactNode;
  /**
   * 辅助性文字介绍
   */
  description?: React.ReactNode;
  /**
   * 是否显示辅助图标
   */
  showIcon?: boolean;
  /**
   * 是否显示关闭按钮，默认为false，即不展示
   */
  closable?: boolean;
  /**
   * 指定提示的样式，四种：info、success、warning、error
   */
  type?: 'info' | 'success' | 'warning' | 'error';
  /**
   * 关闭动画结束后触发的回调函数
   */
  afterClose?: () => void;
  /**
   * 自定义类名
   */
  className?: string;
  /**
   * 子元素
   */
  children?: React.ReactNode;
}

const iconMapFilled = {
  success: AiFillCheckCircle,
  info: AiFillInfoCircle,
  error: AiFillCloseCircle,
  warning: AiFillExclamationCircle,
};

const iconMapOutlined = {
  success: AiOutlineCheckCircle,
  info: AiOutlineInfoCircle,
  error: AiOutlineCloseCircle,
  warning: AiOutlineExclamationCircle,
};

/**
 * Alert组件  警告提示，展现需要关注的信息。
 */
function Alert(props: Props) {
  const [isVisible, setIsVisible] = useState(true);

  const {
    message,
    children,
    description,
    closable,
    type = 'info',
    showIcon,
    className,
    afterClose,
  } = props;

  if (process.env.NODE_ENV === 'development') {
    assert(message || children, 'Alert 组件必须指定 message 或者 children。');
  }

  const handleClose = () => {
    setIsVisible(false);
  };

  const rippleConfig = {
    center: true,
    rippleClassName: 'sinoui-alert-close__ripple',
    rippleLayoutClassName: 'sinoui-alert-close__ripple-layout',
    fixSize: true,
  };

  const closeIcon = closable ? (
    <CloseButton
      onClick={handleClose}
      description={!!description}
      data-testid="closeButton"
      aria-label="Close"
      className="sinoui-alert-close-button"
      ripple={rippleConfig}
    >
      <AiOutlineClose />
    </CloseButton>
  ) : null;

  const IconType = (description ? iconMapOutlined : iconMapFilled)[type];

  return (
    <>
      <CSSTransition
        in={isVisible}
        timeout={timeout}
        classNames="sinoui-alert"
        onExited={afterClose}
        unmountOnExit
      >
        <AlertLayout
          showIcon={showIcon}
          description={!!description}
          closable={closable}
          type={type}
          className={classNames(
            'sinoui-alert',
            className,
            `sinoui-alert--${type}`,
            {
              'sinoui-alert--show-description': !!description,
              'sinoui-alert--closable': closable,
              'sinoui-alert--show-icon': showIcon,
            },
          )}
          role="alert"
        >
          <AlertMessage
            description={!!description}
            data-testid="alertMessage"
            className="sinoui-alert-message"
          >
            {message || children}
          </AlertMessage>
          <AlertDescription className="sinoui-alert-description">
            {description}
          </AlertDescription>
          {closeIcon}
          {showIcon ? (
            <DenseIcon
              description={!!description}
              type={type === 'info' ? 'primary' : type}
              className="sinoui-alert-icon"
            >
              <IconType />
            </DenseIcon>
          ) : null}
        </AlertLayout>
      </CSSTransition>
      <Globalstyle />
    </>
  );
}

export default Alert;
