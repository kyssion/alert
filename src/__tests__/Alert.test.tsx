import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { render, fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { act } from 'react-dom/test-utils';
import TestWrapper from './TestWrapper';
import Alert from '../Alert';

/**
 * Alert 快照测试
 */

describe('Alert 快照测试', () => {
  it('渲染Alert', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <Alert message="提示信息" />
        </TestWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('children测试', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <Alert>提示信息</Alert>
        </TestWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('设置显示关闭按钮', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <Alert message="提示信息" closable />
        </TestWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('设置显示辅助图标', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <Alert message="提示信息" showIcon />
        </TestWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
  it('不同类型的警告提示', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <Alert message="提示信息" showIcon type="success" />
          <Alert message="提示信息" showIcon type="info" />
          <Alert message="提示信息" showIcon type="warning" />
          <Alert message="提示信息" showIcon type="error" />
        </TestWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('设置辅助性文字介绍', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <Alert message="提示信息" description="辅助性文字介绍" showIcon />
        </TestWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('辅助性文字介绍和辅助图标', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <Alert
            message="提示信息"
            description="辅助性文字介绍"
            showIcon
            type="info"
          />
        </TestWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('不同类型辅助性文字介绍和辅助图标', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <Alert
            message="提示信息"
            description="辅助性文字介绍"
            closable
            showIcon
            type="info"
          />
          <Alert
            message="提示信息"
            description="辅助性文字介绍"
            closable
            showIcon
            type="success"
          />
          <Alert
            message="提示信息"
            description="辅助性文字介绍"
            closable
            showIcon
            type="warning"
          />
          <Alert
            message="提示信息"
            description="辅助性文字介绍"
            closable
            showIcon
            type="error"
          />
        </TestWrapper>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('Alert 单元测试', () => {
  // 启用定时器模拟器
  jest.useFakeTimers();

  afterEach(cleanup);

  test('关闭动画结束后触发的回调函数', async () => {
    // 测试首次渲染
    const { getByTestId, container } = render(
      <TestWrapper>
        <Alert
          message="提示信息"
          description="辅助性文字介绍"
          closable
          showIcon
          type="info"
        />
      </TestWrapper>,
    );

    const alertMessage = getByTestId('alertMessage');
    const closeButton = getByTestId('closeButton');

    expect(alertMessage).toHaveTextContent('提示信息');

    // 测试关闭
    act(() => {
      fireEvent.click(closeButton);

      jest.runAllTimers();
    });

    expect(container.querySelector('.sinoui-alert')).toBeFalsy();
  });
});
