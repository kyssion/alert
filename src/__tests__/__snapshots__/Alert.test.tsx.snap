// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert 快照测试 不同类型的警告提示 1`] = `
.c0 {
  position: relative;
  padding: 8px 30px 8px 37px;
  color: rgba(0,0,0,0.65);
  line-height: 1.5;
  border-radius: 4px;
  background-color: #e8f5e9;
  border: 1px solid #c8e6c9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  word-wrap: break-word;
  margin-bottom: 16px;
}

.c2 {
  display: inline-block;
  font-size: 0.875rem;
  margin-bottom: 4px;
  color: rgba(0,0,0,0.85);
}

.c3 {
  font-size: 0.875rem;
  display: inline-block;
  line-height: 22px;
}

.c1 {
  font-family: 'Material Icons';
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  max-width: 1em;
  display: inline-block;
  text-transform: none;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  direction: ltr;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
  color: currentColor;
  -webkit-transition: color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  color: #00c853;
  font-size: 16px;
  margin-right: 8px;
  height: 20px;
  position: absolute;
  top: 10px;
  left: 16px;
}

<div
  className="c0"
  type="success"
>
  <span
    className="c1 sinoui-icon"
    color="success"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm193.5 301.7l-210.6 292a31.8 31.8 0 0 1-51.7 0L318.5 484.9c-3.8-5.3 0-12.7 6.5-12.7h46.9c10.2 0 19.9 4.9 25.9 13.3l71.2 98.8 157.2-218c6-8.3 15.6-13.3 25.9-13.3H699c6.5 0 10.3 7.4 6.5 12.7z"
      />
    </svg>
  </span>
  <div>
    <span
      className="c2"
      data-testid="span"
    >
      提示信息
    </span>
    <span
      className="c3"
    />
  </div>
</div>
`;

exports[`Alert 快照测试 渲染Alert 1`] = `
.c0 {
  position: relative;
  padding: 8px 30px 8px 15px;
  color: rgba(0,0,0,0.65);
  line-height: 1.5;
  border-radius: 4px;
  background-color: #e0f7fa;
  border: 1px solid #b2ebf2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  word-wrap: break-word;
  margin-bottom: 16px;
}

.c1 {
  display: inline-block;
  font-size: 0.875rem;
  margin-bottom: 4px;
  color: rgba(0,0,0,0.85);
}

.c2 {
  font-size: 0.875rem;
  display: inline-block;
  line-height: 22px;
}

<div
  className="c0"
  type="info"
>
  <div>
    <span
      className="c1"
      data-testid="span"
    >
      提示信息
    </span>
    <span
      className="c2"
    />
  </div>
</div>
`;

exports[`Alert 快照测试 设置显示关闭按钮 1`] = `
.c0 {
  position: relative;
  padding: 8px 30px 8px 15px;
  color: rgba(0,0,0,0.65);
  line-height: 1.5;
  border-radius: 4px;
  background-color: #e0f7fa;
  border: 1px solid #b2ebf2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  word-wrap: break-word;
  margin-bottom: 16px;
}

.c2 {
  display: inline-block;
  font-size: 0.875rem;
  margin-bottom: 4px;
  color: rgba(0,0,0,0.85);
}

.c3 {
  font-size: 0.875rem;
  display: inline-block;
  line-height: 22px;
}

.c1 {
  position: absolute;
  top: 10px;
  right: 16px;
  overflow: hidden;
  font-size: 16px;
  line-height: 22px;
  border: 0;
  cursor: pointer;
  background-color: transparent;
  outline: none;
  color: rgba(0,0,0,0.75);
}

<div
  className="c0"
  type="info"
>
  <button
    className="c1"
    data-testid="button"
    onClick={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
      />
    </svg>
  </button>
  <div>
    <span
      className="c2"
      data-testid="span"
    >
      提示信息
    </span>
    <span
      className="c3"
    />
  </div>
</div>
`;

exports[`Alert 快照测试 设置显示辅助图标 1`] = `
.c0 {
  position: relative;
  padding: 8px 30px 8px 37px;
  color: rgba(0,0,0,0.65);
  line-height: 1.5;
  border-radius: 4px;
  background-color: #e0f7fa;
  border: 1px solid #b2ebf2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  word-wrap: break-word;
  margin-bottom: 16px;
}

.c2 {
  display: inline-block;
  font-size: 0.875rem;
  margin-bottom: 4px;
  color: rgba(0,0,0,0.85);
}

.c3 {
  font-size: 0.875rem;
  display: inline-block;
  line-height: 22px;
}

.c1 {
  font-family: 'Material Icons';
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  max-width: 1em;
  display: inline-block;
  text-transform: none;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  direction: ltr;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
  color: currentColor;
  -webkit-transition: color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  color: #00bcd4;
  font-size: 16px;
  margin-right: 8px;
  height: 20px;
  position: absolute;
  top: 10px;
  left: 16px;
}

<div
  className="c0"
  type="info"
>
  <span
    className="c1 sinoui-icon"
    color="info"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm32 664c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V456c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272zm-32-344a48.01 48.01 0 0 1 0-96 48.01 48.01 0 0 1 0 96z"
      />
    </svg>
  </span>
  <div>
    <span
      className="c2"
      data-testid="span"
    >
      提示信息
    </span>
    <span
      className="c3"
    />
  </div>
</div>
`;

exports[`Alert 快照测试 设置显示辅助图标 2`] = `
.c0 {
  position: relative;
  padding: 15px 30px 15px 64px;
  color: rgba(0,0,0,0.65);
  line-height: 1.5;
  border-radius: 4px;
  background-color: #e0f7fa;
  border: 1px solid #b2ebf2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  word-wrap: break-word;
  margin-bottom: 16px;
}

.c2 {
  display: block;
  font-size: 1rem;
  margin-bottom: 4px;
  color: rgba(0,0,0,0.85);
}

.c3 {
  font-size: 0.875rem;
  display: block;
  line-height: 22px;
}

.c1 {
  font-family: 'Material Icons';
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  max-width: 1em;
  display: inline-block;
  text-transform: none;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  direction: ltr;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
  color: currentColor;
  -webkit-transition: color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  color: #00bcd4;
  font-size: 24px;
  margin-right: 8px;
  height: 20px;
  position: absolute;
  top: 14px;
  left: 24px;
}

<div
  className="c0"
  type="info"
>
  <span
    className="c1 sinoui-icon"
    color="info"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
      />
      <path
        d="M464 336a48 48 0 1 0 96 0 48 48 0 1 0-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
      />
    </svg>
  </span>
  <div>
    <span
      className="c2"
      data-testid="span"
    >
      提示信息
    </span>
    <span
      className="c3"
    >
      辅助性文字介绍
    </span>
  </div>
</div>
`;

exports[`Alert 快照测试 设置辅助性文字介绍 1`] = `
.c0 {
  position: relative;
  padding: 15px 30px 15px 64px;
  color: rgba(0,0,0,0.65);
  line-height: 1.5;
  border-radius: 4px;
  background-color: #e0f7fa;
  border: 1px solid #b2ebf2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  word-wrap: break-word;
  margin-bottom: 16px;
}

.c2 {
  display: block;
  font-size: 1rem;
  margin-bottom: 4px;
  color: rgba(0,0,0,0.85);
}

.c3 {
  font-size: 0.875rem;
  display: block;
  line-height: 22px;
}

.c1 {
  font-family: 'Material Icons';
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  max-width: 1em;
  display: inline-block;
  text-transform: none;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  direction: ltr;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
  color: currentColor;
  -webkit-transition: color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  color: #00bcd4;
  font-size: 24px;
  margin-right: 8px;
  height: 20px;
  position: absolute;
  top: 14px;
  left: 24px;
}

<div
  className="c0"
  type="info"
>
  <span
    className="c1 sinoui-icon"
    color="info"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
      />
      <path
        d="M464 336a48 48 0 1 0 96 0 48 48 0 1 0-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
      />
    </svg>
  </span>
  <div>
    <span
      className="c2"
      data-testid="span"
    >
      提示信息
    </span>
    <span
      className="c3"
    >
      辅助性文字介绍
    </span>
  </div>
</div>
`;
